app_integrations/apps/aliyun.py

```diff
@@ -0,0 +1,148 @@
+      "StartTime": <start of the time range of events>,
+      "Events": [entries],
+    }
+
+    If there are no more events in the queried range, the 'NextToken' element is not present.
+
+    Resource:
+        https://www.alibabacloud.com/help/doc-detail/28849.htm
+    """
+
+    _MAX_RESULTS = 50
+
+    def __init__(self, config):
+        super(AliyunApp, self).__init__(config)
+        self._next_token = None
+
+    @classmethod
+    def _type(cls):
+        return 'events'
+
+    @classmethod
+    def service(cls):
+        return 'aliyun'
+
+
+    @classmethod
+    def date_formatter(cls):
+        """Return a format string for a date, ie: 2014-05-26T12:00:00Z
+
+        This format is consisten with the format used by the Aliyun API:
+            https://www.alibabacloud.com/help/doc-detail/28849.htm
+        """
+        return '%Y-%m-%dT%H:%M:%SZ'
+
+    def _gather_logs(self):
+        auth = self._config.auth
+        client = AcsClient(auth['access_key_id'], auth['access_key_secret'], auth['region_id'])
+
+        request = LookupEventsRequest.LookupEventsRequest()
+        request.set_MaxResults(self._MAX_RESULTS)
+        request.set_StartTime = self._last_timestamp
+        if self._next_token:
+            request.set_NextToken(self._next_token)
+
+        try:
+            response = client.do_action_with_exception(request)
+            json_response = json.loads(response)
+            if 'NextToken' in json_response:
+                self._more_to_poll = True
+                self._next_token = json_response['NextToken']
+            else:
+                self._more_to_poll = False
+                self._last_timestamp = json_response['EndTime']
+
+            return json_response['Events']
+
+        except ServerException as e:
+            LOGGER.exception("Server Exception: %s", str(e))
+            return False
+        except ClientException as e:
+            LOGGER.exception("Client Exception: %s", str(e))
+            return False
+
+    @classmethod
+    def _required_auth_info(cls):
+        """Required credentials for access to the resources"""
+
+        def region_validator(region):
+            """Region names pulled from https://www.alibabacloud.com/help/doc-detail/40654.htm"""
+
+            if region in ["cn-qingdao", "cn-beijing", "cn-zhangjiakou", "cn-huhehaote",
+                          "cn-hangzhou", "cn-shanghai", "cn-shenzhen", "cn-hongkong",
+                          "ap-southeast-1", "ap-southeast-2", "ap-southeast-3", "ap-southeast-5",
+                          "ap-northeast-1", "ap-south-1", "us-west-1", "us-east-1",
+                          "eu-central-1", "me-east-1"]:
+                return region
+            return False
+
+        return {
+            'access_key_id': {
+                'description': ('The access key id generated for a RAM user. This '
+                                'should be a string of alphanumeric characters.'),
+                'format': re.compile(r'.*')
```

Comment:
are there no restrictions similar to AWS for their access and secret keys? ie - AWS access keys are 20 alphanumeric all-caps characters and must start with `A`, etc

Reply:
I can't find any documentation anywhere regarding the format. The values I have are a mix of digits and upper and lower case letters, with the id 16 characters in length and the secret 30 characters in length. I'm perfectly willing to put those in as restrictions, but I don't have any reference for that being the supported format.